src/core_plugins/kibana/public/visualize/editor/editor.js

```diff
@@ -316,20 +318,23 @@ function VisEditor($rootScope, $scope, $route, timefilter, AppState, $window, kb
       if (id) {
         notify.info('Saved Visualization "' + savedVis.title + '"');
         if ($scope.isAddToDashMode()) {
-          const savedVisualizationUrl =
-            kbnUrl.eval(`${kbnBaseUrl}#${VisualizeConstants.EDIT_PATH}/{{id}}`, { id: savedVis.id });
-
+          const savedVisualizationParsedUrl = new KibanaParsedUrl({
+            basePath: chrome.getBasePath(),
+            appId: kbnBaseUrl.slice('/app/'.length),
+            appPath: kbnUrl.eval(`${VisualizeConstants.EDIT_PATH}/{{id}}`, { id: savedVis.id })
+          });
           // Manually insert a new url so the back button will open the saved visualization.
-          $window.history.pushState({}, '', `${chrome.getBasePath()}${savedVisualizationUrl}`);
+          $window.history.pushState({}, '', savedVisualizationParsedUrl.getRootRelativePath());
           // Since we aren't reloading the page, only inserting a new browser history item, we need to manually update
           // the last url for this app, so directly clicking on the Visualize tab will also bring the user to the saved
           // url, not the unsaved one.
-          chrome.trackSubUrlForApp('kibana:visualize', savedVisualizationUrl);
+          chrome.trackSubUrlForApp('kibana:visualize', savedVisualizationParsedUrl);
+
+          const lastDashboardAbsoluteUrl = chrome.getNavLinkById('kibana:dashboard').lastSubUrl;
+          const dashboardParsedUrl = absoluteToParsedUrl(lastDashboardAbsoluteUrl, chrome.getBasePath());
+          const appPathWithNewQuery = `${dashboardParsedUrl.appPath}&${DashboardConstants.NEW_VISUALIZATION_ID_PARAM}={{id}}`;
```

Comment:
I suggest using `src/utils/modify_url.js` for this:

```js
import { modifyUrl } from '../../../../../utils';

// ...

kbnUrl.change(modifyUrl(dashboardParsedUrl.appPath, parsed => {
  parsed.query[DashboardConstants.NEW_VISUALIZATION_ID_PARAM] = savedVis.id;
}))
```

It parsed the URL, lets you modify the parsed properties (specifically the [useful bits](https://github.com/elastic/kibana/blob/master/src/utils/modify_url.js#L45-L54) that don't conflict with others), and the formats the URL again to be valid and properly escaped, eliminating the need for `kbnUrl.eval()`.